internal/repository/file.go

```diff
@@ -0,0 +1,311 @@
+		return nil, fmt.Errorf("invalid URI for file: %s", uri)
+	}
+
+	path := uri.String()[7:]
+	var f *os.File
+	var err error
+	if create {
+		f, err = os.Create(path) // If it exists this will truncate which is what we wanted
+	} else {
+		f, err = os.Open(path)
+	}
+	return &file{File: f, uri: uri}, err
+}
+
+// Reader implements repository.Repository.Reader
+//
+// Since: 2.0.0
+func (r *FileRepository) Reader(u fyne.URI) (fyne.URIReadCloser, error) {
+	return openFile(u, false)
+}
+
+// CanRead implements repository.Repository.CanRead
+//
+// Since: 2.0.0
+func (r *FileRepository) CanRead(u fyne.URI) (bool, error) {
+	f, err := os.OpenFile(u.Path(), os.O_RDONLY, 0666)
+	if err == nil {
+		defer f.Close()
+	} else {
+
+		if os.IsPermission(err) {
+			return false, nil
+		}
+
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// Destroy implements repository.Repository.Destroy
+func (r *FileRepository) Destroy(scheme string) {
+	// do nothing
+}
+
+// Writer implements repository.WriteableRepository.Writer
+//
+// Since: 2.0.0
+func (r *FileRepository) Writer(u fyne.URI) (fyne.URIWriteCloser, error) {
+	return openFile(u, true)
+}
+
+// CanWrite implements repository.WriteableRepository.CanWrite
+//
+// Since: 2.0.0
+func (r *FileRepository) CanWrite(u fyne.URI) (bool, error) {
+	f, err := os.OpenFile(u.Path(), os.O_WRONLY, 0666)
+	if err == nil {
+		defer f.Close()
+	} else {
+
+		if os.IsPermission(err) {
+			return false, nil
+		}
+
+		if os.IsNotExist(err) {
+			// We may need to do extra logic to check if the
+			// directory is writeable, but presumably the
+			// IsPermission check covers this.
+			return true, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// Delete implements repository.WriteableRepository.Delete
+//
+// Since: 2.0.0
+func (r *FileRepository) Delete(u fyne.URI) error {
+	return os.Remove(u.Path())
+}
+
+// Parent implements repository.HierarchicalRepository.Parent
+//
+// Since: 2.0.0
+func (r *FileRepository) Parent(u fyne.URI) (fyne.URI, error) {
+	s := u.String()
+
+	// trim trailing slash
+	if s[len(s)-1] == '/' {
+		s = s[0 : len(s)-1]
+	}
+
+	// trim the scheme
+	s = s[len(u.Scheme())+3:]
+
+	// Completely empty URI with just a scheme
+	if len(s) == 0 {
+		return nil, repository.URIRootError
+	}
+
+	parent := ""
+	if u.Scheme() == "file" {
+		// use the system native path resolution
+		parent = filepath.Dir(s)
+		if parent[len(parent)-1] != filepath.Separator {
+			parent += "/"
+		}
+
+		// only root is it's own parent
+		if filepath.Clean(parent) == filepath.Clean(s) {
+			return nil, repository.URIRootError
+		}
+
+		return storage.ParseURI(u.Scheme() + "://" + parent)
+	}
+
+	// Per the spec for storage.Parent(), we should only need to handle
+	// it for schemes we are registered to.
+	return nil, fmt.Errorf("FileRepository cannot handle unknown scheme '%s'", u.Scheme())
```

Comment:
In principle I agree, however this causes a compile error, since the Go compiler does not know that control cannot reach the end of the function without hitting a return, which indeed it could if someone was being naughty.

Reply:
If the function will only ever be called with "file://" URIs then the if statement is not needed

Reply:
Ahh, good point. ebedbf7d2e76ee0a2cdfc2705db85b779915eeb9